main.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"checkout_system/pkg/utils"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+
+	"github.com/joho/godotenv"
+	"github.com/sirupsen/logrus"
+)
+
+type Products struct {
+	Products []Product `json:"products"`
+}
+
+type Product struct {
+	SKU   string  `json:"SKU"`
+	Name  string  `json:"Name"`
+	Price float32 `json:"Price"`
+}
+
+func (p *Products) ReadProducts() {
+	log := logrus.New().WithField("function", "ReadProducts()")
+
+	inputFile, err := os.Open(os.Getenv("INPUT_FILENAME"))
+	if err != nil {
+		log.Error("invalid path: ", err)
+		return
+	}
+
+	data, err := ioutil.ReadAll(inputFile)
+	if err != nil {
+		log.Error("cannot read input file: ", err)
+		return
+	}
+
+	defer func(inputFile *os.File) {
+		err := inputFile.Close()
+		if err != nil {
+			log.Errorf("failed to close `%s`", inputFile.Name())
+		}
+	}(inputFile)
+
+	err = json.Unmarshal(data, &p)
+	if err != nil {
+		log.Error("unmarshalling error: ", err)
+
+	}
+	log.Debug("successfully read %d products", len(p.Products))
+}
+
+func (p Product) WithDropSpecial(cnt int) {
```

Comment:
What if we simplify interface of all functions like that? 

For example, if we modify it to the following: 

```
package offers 

func ApplySuperIpadPrice(cart *cart.ProductCart) *cart.ProductCart {
   if  cart.Contains("ipd") && cart.Get("ipd").Len() > 4 {
     cart.UpdateItemPrice("ipd", 499.99)
   }
   
   return cart
}
```

Using this pattern we can easily combine our special offers, e.g.: 

```
myShoppingCart := cart.New()

myShoppingCart.Add("ipd")
myShoppingCart.Add("...")
... 

specialOffers := []func(*cart.ProductCart) *cart.ProductCart {
   offers.ApplyVGASuperDeal, 
   offers.ApplyAppleTVSpecialPrice,
   offers.ApplySuperIpadPrice
}

for _, o := range specialOffers {
    myShoppingCart = o(myShoppingCart)
}

fmt.Println(myShoppingCart.Total())
```